src/hotspot/share/opto/mulnode.cpp

```diff
@@ -369,17 +369,24 @@ Node *MulLNode::Ideal(PhaseGVN *phase, bool can_reshape) {
   if (bit1 == abs_con) {           // Found a power of 2?
     res = new LShiftLNode(in(1), phase->intcon(log2i_exact(bit1)));
   } else {
-
     // Check for constant with 2 bits set
     julong bit2 = abs_con-bit1;
     bit2 = bit2 & (0-bit2);          // Extract 2nd bit
     if (bit2 + bit1 == abs_con) {    // Found all bits in con?
+      if (!phase->C->post_loop_opts_phase()) {
```

Comment:
Thanks or adding the MulL case.  Nowadays the range check optimizations don't apply only to arrays, and apply to both int and long indexes, mainly because of Panama.